packages/product/src/services/product-variant.ts

```diff
@@ -23,6 +24,30 @@ export default class ProductVariantService<
     this.productService_ = productService
   }
 
+  async retrieve(
+    productVariantId: string,
+    config: FindConfig<ProductTypes.ProductVariantDTO> = {},
+    sharedContext?: Context
+  ): Promise<TEntity> {
+    const queryOptions = ModulesSdkUtils.buildQuery<TEntity>({
```

Comment:
suggestion: You can use ProductVariant here I think since TEntity must extends at least ProductVariant. wdyt? I am fine keeping it as it is but I wanted to get your feedback.

Reply:
Curious about this convo. Why are we not typing this method with `ProductVariant`?

Reply:
And other methods

Reply:
This is something I'm still wrapping my head around.

But the discussion stemmed from the model and repository layer being completely replaceable for which the service should not be tied to model and repository types. There is a public interface that it must obey for repositories `DAL.RepositoryService`, but we don't have something similar for models.

If we are passing ProductVariant everywhere, then I don't understand the role of TEntity here and why we can't just use ProductVariant in each case as we are monkey patching them anyway.

Reply:
@riqwan the explanation is the exact same as for the repository, which we have discussed again last day 😊

TEntity allows the users to specify which type it will be based on for the input/output, not for us. In short, TEntity extends ProdfuctVariant, therefore TEntity at least contains the properties of product variant but can be instanciated with another type than product variant. If we didn't allow the user to customise the DAL, it wouldn't be necessary, but the fact is that internally we know that tentity is at least product variant, but if the users customise it, then it will be able to return more than product variant, it is for that reason that we can use the entity as the type here while returning TEntity for the user that consume it because of that constraints that the type should extends ProductVariant. 

If it is not clear, we can re discuss that tomorrow wdyt?

Reply:
Long story short (summary): we know that TEntity at least extends ProductVariant, we are based on ProductVariant, if a user extends ProductVariant, he can pass it as TEntity which will be the type on whoch the input/output will be based on. Without TEntity, the user will never be able to specify that it is in fact more than ProductVariant and will have to go through a definition file to extend the original type like we explain to do in the core. Don't know if it is shorter 🤣

Reply:
So, given that explanation, which one would you say we should type `buildQuery` with?

Reply:
I would the entity, and I need to do the changes on my side as well I think

Reply:
Meaning, TEntity right? Then we're aligned. 💪🏻

Reply:
no meaning the proper entity :)

Reply:
if we don't touch the ouput of build query then we can use TEntity otherwise we have to use the entity, so I would say lets have only one thing instead of a mix of both

Reply:
I think it's evident that this needs documenting with very clear examples of both types of usage. @adrien2p can I get you to draft a couple of sentences about this?

Reply:
TEntity will be used for external public API, input/output types so that any users can provide this as long as it extends the Entity. Internally, we are using the entity itself to have proper typings as we know that TEntity is at least the entity and an extends constraint is present. We can see that in the template itself where we say that the default value for TEntity is the entity itself